### PR TITLE
Feedback

### PR DESCRIPTION
**Pull request by Florisheinen1**
RobotHub _does_ receive feedback from both the Simulator and the Basestation, but _does not_ forward it. This PR converts the feedback from the simulator and feedback from the basestation into messages that can be forwarded.

### CODE REVIEW

File: src/simulation/SimulatorManager.cpp [hunk 1]
```diff
@@ -101,6 +102,7 @@ void SimulatorManager::listenForConfigurationFeedback() {
                 this->callConfigurationFeedbackCallback(f);
             }
         }
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (alexander-de-ranitz) [inline src/simulation/SimulatorManager.cpp]:
```suggestion
        std::this_thread::sleep_for(std::chrono::milliseconds(LISTEN_THREAD_COOLDOWN_MS));
```

File: src/simulation/SimulatorManager.cpp [hunk 2]
```diff
@@ -89,6 +89,7 @@ void SimulatorManager::listenForRobotControlFeedback(bool listenForTeamYellow) {
                 this->callRobotControlFeedbackCallback(f);
             }
         }
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (alexander-de-ranitz) [inline src/simulation/SimulatorManager.cpp]:
```suggestion
        std::this_thread::sleep_for(std::chrono::milliseconds(LISTEN_THREAD_COOLDOWN_MS));
```

Comment (alexander-de-ranitz) [reply src/simulation/SimulatorManager.cpp]: might need to #include constants.h

File: src/RobotHub.cpp [hunk 3]
```diff
@@ -119,13 +114,38 @@ void RobotHub::printStatistics() {
...
 }
-void handleRobotFeedbackFromBasestation(const RobotFeedback &feedback) {
-    // std::cout << "Received robot feedback from the basestation!" << std::endl;
-    // TODO: Forward feedback to AI or something idk
+
+void RobotHub::handleRobotFeedbackFromBasestation(const RobotFeedback &feedback) {
+    proto::RobotData feedbackToBePublished;
+    feedbackToBePublished.set_isyellow(this->settings.isyellow());
+
+    // TODO: Perhaps wait for all robots to return feedback to publish combined feedback message for efficiency
```
Comment (emielsteerneman) [inline src/RobotHub.cpp]: No. Feedback might get lost in the wireless transmission.

File: src/RobotHub.cpp [hunk 4]
```diff
@@ -119,13 +114,38 @@ void RobotHub::printStatistics() {
...
+    proto::RobotFeedback *feedbackOfRobot = feedbackToBePublished.add_receivedfeedback();
+    feedbackOfRobot->set_id(feedback.id);
+    feedbackOfRobot->set_xsenscalibrated(feedback.XsensCalibrated);
+    feedbackOfRobot->set_ballsensorisworking(feedback.ballSensorWorking);
+    feedbackOfRobot->set_batterylow(feedback.batteryLevel <= BATTERY_LOW_LEVEL);
+    feedbackOfRobot->set_hasball(feedback.hasBall);
+    feedbackOfRobot->set_ballpos(feedback.ballPos);
+    // feedbackOfRobot->set_x_vel(feedback.)
```
Comment (emielsteerneman) [inline src/RobotHub.cpp]: ?

File: src/basestation/BasestationManager.cpp [hunk 5]
```diff
@@ -64,7 +63,7 @@ bool BasestationManager::sendSerialCommand(RobotCommandPayload &payload) const {
  *  receives feedback. If the callback function is not set, no feedback will be sent to AI.
  *  @param callback The function that will be called whenever feedback is received
  */
-void BasestationManager::setFeedbackCallback(const std::function<void(RobotFeedback &)> &callback) { this->feedbackCallbackFunction = callback; }
+void BasestationManager::setFeedbackCallback(const std::function<void(const RobotFeedback &)> &callback) { this->feedbackCallbackFunction = callback; }
```
Comment (emielsteerneman) [inline src/basestation/BasestationManager.cpp:66]: This doesn't seem in line with all the formatting that's been going on lately

File: include/simulation/Feedback.hpp [hunk 6]
```diff
@@ -11,7 +12,7 @@ typedef struct SimulationError {
 
 typedef struct RobotControlFeedback {
     bool isTeamYellow;
-    std::vector<int> robotsThatHaveTheBall;
+    std::map<int, bool> robotIdHasBall;
```
Comment (emielsteerneman) [inline include/simulation/Feedback.hpp:15]: Is this map ever cleared? If a robot is taken out of the field, is it also taken out of the map?

File: src/RobotHub.cpp [hunk 7]
```diff
@@ -119,13 +114,38 @@ void RobotHub::printStatistics() {
...
 
-void handleRobotFeedbackFromSimulator(const simulation::RobotControlFeedback &feedback) {
-    // std::cout << "Received robot feedback from the simulator!" << std::endl;
-    // TODO: Forward feedback to AI or something idk
+void RobotHub::handleRobotFeedbackFromSimulator(const simulation::RobotControlFeedback &feedback) {
+    proto::RobotData feedbackToBePublished;
+    feedbackToBePublished.set_isyellow(feedback.isTeamYellow);
+
+    for (auto const &[robotId, hasBall] : feedback.robotIdHasBall) {
+        proto::RobotFeedback *feedbackOfRobot = feedbackToBePublished.add_receivedfeedback();
```
Comment (emielsteerneman) [inline src/RobotHub.cpp:123]: Should this not also be `proto::RobotData` instead of `proto::RobotFeedback` ?

**Review by emielsteerneman — COMMENTED** [review]
lgtm, as long as its tested with the actual robots. The only concern I have is with `robotIdHasBall` possibly not being cleared.

**Review by alexander-de-ranitz — APPROVED** [review]
lgmt

**Pull request merged**